File: dnstwister/dnstwist/dnstwist.py
# -*- coding: utf-8 -*-
#
# dnstwist
#
# Generate and resolve domain variations to detect typo squatting,
# phishing and corporate espionage.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#
# dnstwist version: https://github.com/elceef/dnstwist/blob/182902f42c749cc4b58af06f8c312c92af1a73dc/dnstwist.py
# modified by Robert Wallhead (robert@thisismyrobot.com) for use in
# https://dnstwister.report - all functionality except fuzzing removed and
# various changes made to support usage in Heroku.
#

__author__ = 'Marcin Ulikowski'
__version__ = '20180623'
__email__ = 'marcin@ulikowski.pl'

import re
import os.path

import idna


FILE_TLD = os.path.join(
    'dnstwister',
    'dnstwist',
    'database',
    'effective_tld_names.dat'
)
DB_TLD = os.path.exists(FILE_TLD)
if not DB_TLD:
    raise Exception('TLD database is required!')

VALID_DOMAIN_RE = re.compile(
    r'(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}\.?$)',
    flags=re.IGNORECASE
)


class InvalidDomain(Exception):
    """ Exception for invalid domains.
    """
    pass


class Result(object):

    def __init__(self, fuzzer, domain):
        self._fuzzer = fuzzer
        self._domain = domain

    @property
    def fuzzer(self):
        return self._fuzzer

    @property
    def domain(self):
        return self._domain


class ResultBuilder(object):

    def __init__(self, tld):
        self._tld = tld

    def build(self, fuzzer, domain):
        return Result(fuzzer, domain + '.' + self._tld)


def is_valid_domain(domain):
    """Validate a domain - including unicode domains."""
    try:
        if len(domain) > 255:
            return False

        encoded_domain = idna.encode(domain)
        if domain != encoded_domain and len(domain) == encoded_domain:
            return False

        return VALID_DOMAIN_RE.match(encoded_domain) is not None
    except (UnicodeError, TypeError, idna.IDNAError):
        pass
    return False


class fuzz_domain(object):
    """ Domain fuzzer.
    """
    def __init__(self, domain):
        self.domain, self.tld = self.__domain_tld(domain)
        self.domains = []
        self.qwerty = {
        '1': '2q', '2': '3wq1', '3': '4ew2', '4': '5re3', '5': '6tr4', '6': '7yt5', '7': '8uy6', '8': '9iu7', '9': '0oi8', '0': 'po9',
        'q': '12wa', 'w': '3esaq2', 'e': '4rdsw3', 'r': '5tfde4', 't': '6ygfr5', 'y': '7uhgt6', 'u': '8ijhy7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0',
        'a': 'qwsz', 's': 'edxzaw', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'yhbvft', 'h': 'ujnbgy', 'j': 'ikmnhu', 'k': 'olmji', 'l': 'kop',
        'z': 'asx', 'x': 'zsdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhjm', 'm': 'njk'
        }
        self.qwertz = {
        '1': '2q', '2': '3wq1', '3': '4ew2', '4': '5re3', '5': '6tr4', '6': '7zt5', '7': '8uz6', '8': '9iu7', '9': '0oi8', '0': 'po9',
        'q': '12wa', 'w': '3esaq2', 'e': '4rdsw3', 'r': '5tfde4', 't': '6zgfr5', 'z': '7uhgt6', 'u': '8ijhz7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0',
        'a': 'qwsy', 's': 'edxyaw', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'zhbvft', 'h': 'ujnbgz', 'j': 'ikmnhu', 'k': 'olmji', 'l': 'kop',
        'y': 'asx', 'x': 'ysdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhjm', 'm': 'njk'
        }
        self.azerty = {
        '1': '2a', '2': '3za1', '3': '4ez2', '4': '5re3', '5': '6tr4', '6': '7yt5', '7': '8uy6', '8': '9iu7', '9': '0oi8', '0': 'po9',
        'a': '2zq1', 'z': '3esqa2', 'e': '4rdsz3', 'r': '5tfde4', 't': '6ygfr5', 'y': '7uhgt6', 'u': '8ijhy7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0m',
        'q': 'zswa', 's': 'edxwqz', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'yhbvft', 'h': 'ujnbgy', 'j': 'iknhu', 'k': 'olji', 'l': 'kopm', 'm': 'lp',
        'w': 'sxq', 'x': 'zsdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhj'
        }
        self.keyboards = [ self.qwerty, self.qwertz, self.azerty ]

    def __domain_tld(self, domain):
        domain = domain.rsplit('.', 2)

        if len(domain) == 2:
            return domain[0], domain[1]

        if DB_TLD:
            cc_tld = {}
            re_tld = re.compile(r'^[a-z]{2,4}\.[a-z]{2}$', re.IGNORECASE)

            for line in open(FILE_TLD):
                line = line[:-1]
                if re_tld.match(line):
                    sld, tld = line.split('.')
                    if not tld in cc_tld:
                        cc_tld[tld] = []
                    cc_tld[tld].append(sld)

            sld_tld = cc_tld.get(domain[2])
            if sld_tld:
                if domain[1] in sld_tld:
                    return domain[0], domain[1] + '.' + domain[2]

        return domain[0] + '.' + domain[1], domain[2]

    def __validate_domain(self, domain):
        return is_valid_domain(domain)

    def __filter_domains(self):

        # IDNA encoding's detailed check makes this 4x slower, and we validate
        # all requests that just query a domain later on.
        old_func = idna.core.check_label
        idna.core.check_label = lambda l: None

        seen = set()
        filtered = []

        for d in self.domains:
            if d['domain-name'] in seen:
                continue

            seen.add(d['domain-name'])

            if self.__validate_domain(d['domain-name']):
                filtered.append(d)

        self.domains = filtered

        idna.core.check_label = old_func

    def __bitsquatting(self):
        masks = [1, 2, 4, 8, 16, 32, 64, 128]
        for i in range(0, len(self.domain)):
            c = self.domain[i]
            for j in range(0, len(masks)):

                # Deal with Unicode later...
                if ord(c) > 255:
                    continue

                b = chr(ord(c) ^ masks[j])
                o = ord(b)
                if (o >= 48 and o <= 57) or (o >= 97 and o <= 122) or o == 45:
                    yield self.domain[:i] + b + self.domain[i+1:]

    def __homoglyph(self, MAX=1000):
        glyphs = {
        'a': [u'à', u'á', u'â', u'ã', u'ä', u'å', u'ɑ', u'а', u'ạ', u'ǎ', u'ă', u'ȧ', u'ӓ'],
        'b': ['d', 'lb', 'ib', u'ʙ', u'Ь', u'b̔', u'ɓ', u'Б'],
        'c': [u'ϲ', u'с', u'ƈ', u'ċ', u'ć', u'ç'],
        'd': ['b', 'cl', 'dl', 'di', u'ԁ', u'ժ', u'ɗ', u'đ'],
        'e': [u'é', u'ê', u'ë', u'ē', u'ĕ', u'ě', u'ė', u'е', u'ẹ', u'ę', u'є', u'ϵ', u'ҽ'],
        'f': [u'Ϝ', u'ƒ', u'Ғ'],
        'g': ['q', u'ɢ', u'ɡ', u'Ԍ', u'Ԍ', u'ġ', u'ğ', u'ց', u'ǵ', u'ģ'],
        'h': ['lh', 'ih', u'һ', u'հ', u'Ꮒ', u'н'],
        'i': ['1', 'l', u'Ꭵ', u'í', u'ï', u'ı', u'ɩ', u'ι', u'ꙇ', u'ǐ', u'ĭ', u'ì'],
        'j': [u'ј', u'ʝ', u'ϳ', u'ɉ'],
        'k': ['lk', 'ik', 'lc', u'κ', u'ⲕ', u'κ'],
        'l': ['1', 'i', u'ɫ', u'ł'],
        'm': ['n', 'nn', 'rn', 'rr', u'ṃ', u'ᴍ', u'м', u'ɱ'],
        'n': ['m', 'r', u'ń'],
        'o': ['0', u'Ο', u'ο', u'О', u'о', u'Օ', u'ȯ', u'ọ', u'ỏ', u'ơ', u'ó', u'ö', u'ӧ'],
        'p': [u'ρ', u'р', u'ƿ', u'Ϸ', u'Þ'],
        'q': ['g', u'զ', u'ԛ', u'գ', u'ʠ'],
        'r': [u'ʀ', u'Г', u'ᴦ', u'ɼ', u'ɽ'],
        's': [u'Ⴝ', u'Ꮪ', u'ʂ', u'ś', u'ѕ'],
        't': [u'τ', u'т', u'ţ'],
        'u': [u'μ', u'υ', u'Ս', u'ս', u'ц', u'ᴜ', u'ǔ', u'ŭ'],
        'v': [u'ѵ', u'ν', u'v̇'],
        'w': ['vv', u'ѡ', u'ա', u'ԝ'],
        'x': [u'х', u'ҳ', u'ẋ'],
        'y': [u'ʏ', u'γ', u'у', u'Ү', u'ý'],
        'z': [u'ʐ', u'ż', u'ź', u'ʐ', u'ᴢ']
        }

        found = 0

        for ws in range(0, len(self.domain)):
            for i in range(0, (len(self.domain)-ws)+1):
                win = self.domain[i:i+ws]

                j = 0
                while j < ws:
                    c = win[j]
                    if c in glyphs:
                        win_copy = win
                        for g in glyphs[c]:
                            win = win.replace(c, g)
                            yield self.domain[:i] + win + self.domain[i+ws:]
                            win = win_copy

                            # Very long domains have terrible complexity when
                            # ran through this algorithm.
                            found += 1
                            if found >= MAX:
                                return
                    j += 1

    def __hyphenation(self):
        result = []

        for i in range(1, len(self.domain)):
            result.append(self.domain[:i] + '-' + self.domain[i:])

        return result

    def __insertion(self):
        result = set()

        for i in range(1, len(self.domain)-1):
            for keys in self.keyboards:
                if self.domain[i] in keys:
                    for c in keys[self.domain[i]]:
                        result.add(self.domain[:i] + c + self.domain[i] + self.domain[i+1:])
                        result.add(self.domain[:i] + self.domain[i] + c + self.domain[i+1:])

        return result

    def __omission(self):
        result = set()

        for i in range(0, len(self.domain)):
            result.add(self.domain[:i] + self.domain[i+1:])

        n = re.sub(r'(.)\1+', r'\1', self.domain)

        if n not in result and n != self.domain:
            result.add(n)

        return result

    def __repetition(self):
        result = set()

        for i in range(0, len(self.domain)):
            if self.domain[i].isalpha():
                result.add(self.domain[:i] + self.domain[i] + self.domain[i] + self.domain[i+1:])

        return result

    def __replacement(self):
        result = set()

        for i in range(0, len(self.domain)):
            for keys in self.keyboards:
                if self.domain[i] in keys:
                    for c in keys[self.domain[i]]:
                        result.add(self.domain[:i] + c + self.domain[i+1:])

        return result

    def __subdomain(self):
        result = []

        for i in range(1, len(self.domain)):
            if self.domain[i] not in ['-', '.'] and self.domain[i-1] not in ['-', '.']:
                result.append(self.domain[:i] + '.' + self.domain[i:])

        return result

    def __transposition(self):
        result = []

        for i in range(0, len(self.domain)-1):
            if self.domain[i+1] != self.domain[i]:
                result.append(self.domain[:i] + self.domain[i+1] + self.domain[i] + self.domain[i+2:])

        return result

    def __vowel_swap(self):
        vowels = 'aeiou'
        result = set()

        for i in range(0, len(self.domain)):
            for vowel in vowels:
                if self.domain[i] in vowels:
                    result.add(self.domain[:i] + vowel + self.domain[i+1:])

        return result

    def __addition(self):
        for i in range(97, 123):
            yield self.domain + chr(i)

    def fuzz(self):
        """ Perform a domain fuzz.
        """
        self.domains.append({ 'fuzzer': 'Original*', 'domain-name': self.domain + '.' + self.tld })

        for domain in self.__addition():
            self.domains.append({ 'fuzzer': 'Addition', 'domain-name': domain + '.' + self.tld })
        for domain in self.__bitsquatting():
            self.domains.append({ 'fuzzer': 'Bitsquatting', 'domain-name': domain + '.' + self.tld })
        for domain in self.__homoglyph():
            self.domains.append({ 'fuzzer': 'Homoglyph', 'domain-name': domain + '.' + self.tld })
        for domain in self.__hyphenation():
            self.domains.append({ 'fuzzer': 'Hyphenation', 'domain-name': domain + '.' + self.tld })
        for domain in self.__insertion():
            self.domains.append({ 'fuzzer': 'Insertion', 'domain-name': domain + '.' + self.tld })
        for domain in self.__omission():
            self.domains.append({ 'fuzzer': 'Omission', 'domain-name': domain + '.' + self.tld })
        for domain in self.__repetition():
            self.domains.append({ 'fuzzer': 'Repetition', 'domain-name': domain + '.' + self.tld })
        for domain in self.__replacement():
            self.domains.append({ 'fuzzer': 'Replacement', 'domain-name': domain + '.' + self.tld })
        for domain in self.__subdomain():
            self.domains.append({ 'fuzzer': 'Subdomain', 'domain-name': domain + '.' + self.tld })
        for domain in self.__transposition():
            self.domains.append({ 'fuzzer': 'Transposition', 'domain-name': domain + '.' + self.tld })
        for domain in self.__vowel_swap():
            self.domains.append({ 'fuzzer': 'Vowel swap', 'domain-name': domain + '.' + self.tld })

        if not self.domain.startswith('www.'):
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': 'ww' + self.domain + '.' + self.tld })
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': 'www' + self.domain + '.' + self.tld })
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': 'www-' + self.domain + '.' + self.tld })
        if '.' in self.tld:
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': self.domain + '.' + self.tld.split('.')[-1] })
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': self.domain + self.tld })
        if '.' not in self.tld:
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': self.domain + self.tld + '.' + self.tld })
        if self.tld != 'com' and '.' not in self.tld:
            self.domains.append({ 'fuzzer': 'Various', 'domain-name': self.domain + '-' + self.tld + '.com' })

        self.__filter_domains()

    def fuzz_iter(self, de_dupe=False):
        """Return an iterator of the fuzz.

        The intent is to reduce memory usage and to allow the fuzzed domains
        to be returned in a chunked manner over HTTP chunking to the
        front-end.

        The sacrifice of some performance should be lost in the time taken to
        individually resolve each domain - aka an additional 0.001 sec per
        domain here is irrelevant if it takes 1 second to resolve each one
        in the browser.

        You can optionally de-duplicate as you go, though that will use more
        memory obviously.
        """
        seen = set()
        builder = ResultBuilder(self.tld)

        yield builder.build('Original*', self.domain)

        fuzzers = {
            'Addition': self.__addition,
            'Bitsquatting': self.__bitsquatting,
            'Homoglyph': lambda: self.__homoglyph(100000)
        }

        for (tag, fuzzer_func) in fuzzers.items():

            for domain in fuzzer_func():
                if de_dupe:
                    if domain in seen:
                        continue
                    else:
                        seen.add(domain)

                if not is_valid_domain(domain + '.' + self.tld):
                    continue

                yield builder.build(tag, domain)

File: dnstwister/static/sources/dnstwist.js
/* globals tldjs */
var dnstwistjs = (function () {
  // https://wsvincent.com/javascript-array-range-function/
  var range = function (start, edge, step) {
    // If only 1 number passed make it the edge and 0 the start
    if (arguments.length === 1) {
      edge = start
      start = 0
    }

    // Validate edge/start
    edge = edge || 0
    step = step || 1

    // Create array of numbers, stopping before the edge
    var arr = []
    for (arr; (edge - start) * step > 0; start += step) {
      arr.push(start)
    }
    return arr
  }

  var longestPropertyLength = function (obj) {
    var longestLength = 0
    for (var srcChar in obj) {
      var length = obj[srcChar].length
      if (length > longestLength) {
        longestLength = length
      }
    }
    return longestLength
  }

  var homoglyphs = {
    'a': ['à', 'á', 'â', 'ã', 'ä', 'å', 'ɑ', 'а', 'ạ', 'ǎ', 'ă', 'ȧ', 'ӓ'],
    'b': ['d', 'lb', 'ib', 'ʙ', 'Ь', 'b̔', 'ɓ', 'Б'],
    'c': ['ϲ', 'с', 'ƈ', 'ċ', 'ć', 'ç'],
    'd': ['b', 'cl', 'dl', 'di', 'ԁ', 'ժ', 'ɗ', 'đ'],
    'e': ['é', 'ê', 'ë', 'ē', 'ĕ', 'ě', 'ė', 'е', 'ẹ', 'ę', 'є', 'ϵ', 'ҽ'],
    'f': ['Ϝ', 'ƒ', 'Ғ'],
    'g': ['q', 'ɢ', 'ɡ', 'Ԍ', 'Ԍ', 'ġ', 'ğ', 'ց', 'ǵ', 'ģ'],
    'h': ['lh', 'ih', 'һ', 'հ', 'Ꮒ', 'н'],
    'i': ['1', 'l', 'Ꭵ', 'í', 'ï', 'ı', 'ɩ', 'ι', 'ꙇ', 'ǐ', 'ĭ', 'ì'],
    'j': ['ј', 'ʝ', 'ϳ', 'ɉ'],
    'k': ['lk', 'ik', 'lc', 'κ', 'ⲕ', 'κ', '𝖐', '𝓴', '𝚔', '𝔨', '𝒌', '𝘬', '𝓀', '𝙠', '𝐤', '𝗄', '𝑘', '𝗸', '𝕜'],
    'l': ['1', 'i', 'ɫ', 'ł'],
    'm': ['n', 'nn', 'rn', 'rr', 'ṃ', 'ᴍ', 'м', 'ɱ'],
    'n': ['m', 'r', 'ń'],
    'o': ['0', 'ο', 'о', 'Օ', 'ȯ', 'ọ', 'ỏ', 'ơ', 'ó', 'ö', 'ӧ'],
    'p': ['ρ', 'р', 'ƿ', 'Ϸ', 'Þ'],
    'q': ['g', 'զ', 'ԛ', 'գ', 'ʠ'],
    'r': ['ʀ', 'Г', 'ᴦ', 'ɼ', 'ɽ'],
    's': ['Ⴝ', 'Ꮪ', 'ʂ', 'ś', 'ѕ'],
    't': ['τ', 'т', 'ţ'],
    'u': ['μ', 'υ', 'Ս', 'ս', 'ц', 'ᴜ', 'ǔ', 'ŭ'],
    'v': ['ѵ', 'ν', 'v̇'],
    'w': ['vv', 'ѡ', 'ա', 'ԝ'],
    'x': ['х', 'ҳ', 'ẋ'],
    'y': ['ʏ', 'γ', 'у', 'Ү', 'ý'],
    'z': ['ʐ', 'ż', 'ź', 'ʐ', 'ᴢ']
  }
  var longestHomoglyph = longestPropertyLength(homoglyphs)

  var keyboardLayouts = [
    // qwerty
    {
      '1': '2q', '2': '3wq1', '3': '4ew2', '4': '5re3', '5': '6tr4', '6': '7yt5', '7': '8uy6', '8': '9iu7', '9': '0oi8', '0': 'po9',
      'q': '12wa', 'w': '3esaq2', 'e': '4rdsw3', 'r': '5tfde4', 't': '6ygfr5', 'y': '7uhgt6', 'u': '8ijhy7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0',
      'a': 'qwsz', 's': 'edxzaw', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'yhbvft', 'h': 'ujnbgy', 'j': 'ikmnhu', 'k': 'olmji', 'l': 'kop',
      'z': 'asx', 'x': 'zsdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhjm', 'm': 'njk'
    },
    // qwertz
    {
      '1': '2q', '2': '3wq1', '3': '4ew2', '4': '5re3', '5': '6tr4', '6': '7zt5', '7': '8uz6', '8': '9iu7', '9': '0oi8', '0': 'po9',
      'q': '12wa', 'w': '3esaq2', 'e': '4rdsw3', 'r': '5tfde4', 't': '6zgfr5', 'z': '7uhgt6', 'u': '8ijhz7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0',
      'a': 'qwsy', 's': 'edxyaw', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'zhbvft', 'h': 'ujnbgz', 'j': 'ikmnhu', 'k': 'olmji', 'l': 'kop',
      'y': 'asx', 'x': 'ysdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhjm', 'm': 'njk'
    },
    // azerty
    {
      '1': '2a', '2': '3za1', '3': '4ez2', '4': '5re3', '5': '6tr4', '6': '7yt5', '7': '8uy6', '8': '9iu7', '9': '0oi8', '0': 'po9',
      'a': '2zq1', 'z': '3esqa2', 'e': '4rdsz3', 'r': '5tfde4', 't': '6ygfr5', 'y': '7uhgt6', 'u': '8ijhy7', 'i': '9okju8', 'o': '0plki9', 'p': 'lo0m',
      'q': 'zswa', 's': 'edxwqz', 'd': 'rfcxse', 'f': 'tgvcdr', 'g': 'yhbvft', 'h': 'ujnbgy', 'j': 'iknhu', 'k': 'olji', 'l': 'kopm', 'm': 'lp',
      'w': 'sxq', 'x': 'wsdc', 'c': 'xdfv', 'v': 'cfgb', 'b': 'vghn', 'n': 'bhj'
    }
  ]
  var totalLayouts = keyboardLayouts.length
  var longestSet = Math.max.apply(null, keyboardLayouts.map(function (kl) {
    return longestPropertyLength(kl)
  }))

  var vowels = 'aeiou'

  var tweaks = [
  ].concat(

    // Original
    function (d, i) {
      if (i === 0) {
        return d
      }
      return null
    }

  ).concat(

    // Addition
    range(97, 123).map(function (c) {
      return function (d, i) {
        if (i === d.length - 1) {
          return d + String.fromCharCode(c)
        }
        return null
      }
    })

  ).concat(

    // Bitsquatting
    range(8).map(function (m) {
      var mask = Math.pow(2, m)

      return function (d, i) {
        var newChar = d.charCodeAt(i) ^ mask
        if ((newChar >= 48 && newChar <= 57) || (newChar >= 97 && newChar <= 122) || newChar === 45) {
          return d.substr(0, i) + String.fromCharCode(newChar) + d.substr(i + 1)
        } else {
          return null
        }
      }
    })

  ).concat(

    // Homoglyph
    range(longestHomoglyph).map(function (m) {
      return function (d, i) {
        var h = homoglyphs[d[i]]
        if (h === undefined) {
          return null
        }

        var newChar = h[m]
        if (newChar === undefined) {
          return null
        }

        return d.substr(0, i) + newChar + d.substr(i + 1)
      }
    })

  ).concat(

    // Hyphenation
    function (d, i) {
      if (i === 0 || i === d.length) {
        return null
      }
      if (d[i] === '.' || d[i - 1] === '.' || d[i] === '-') {
        return null
      }
      return d.substr(0, i) + '-' + d.substr(i)
    }

  ).concat(

    // Insertion
    range(totalLayouts * longestSet * 2).map(function (l) {
      return function (d, i) {
        if (i === 0) {
          return null
        }

        var layoutIndex = l % totalLayouts
        var charIndex = Math.floor((l / totalLayouts) / 2)

        var keyboardSet = keyboardLayouts[layoutIndex][d[i]]
        if (keyboardSet === undefined) {
          return null
        }

        var insertedChar = keyboardSet[charIndex]
        if (insertedChar === undefined) {
          return null
        }

        if (l % 2 === 0) {
          return d.substr(0, i) + insertedChar + d[i] + d.substr(i + 1)
        } else {
          return d.substr(0, i) + d[i] + insertedChar + d.substr(i + 1)
        }
      }
    })

  ).concat(

    // Omission
    function (d, i) {
      return d.substr(0, i) + d.substr(i + 1)
    }

  ).concat(

    // Repetition
    function (d, i) {
      return d.substr(0, i) + d[i] + d.substr(i)
    }

  ).concat(

    // Replacement
    range(totalLayouts * longestSet).map(function (l) {
      return function (d, i) {
        var layoutIndex = l % totalLayouts
        var charIndex = Math.floor(l / totalLayouts)

        var keyboardSet = keyboardLayouts[layoutIndex][d[i]]
        if (keyboardSet === undefined) {
          return null
        }

        var replacedChar = keyboardSet[charIndex]
        if (replacedChar === undefined) {
          return null
        }

        return d.substr(0, i) + replacedChar + d.substr(i + 1)
      }
    })

  ).concat(

    // Sub-domain
    function (d, i) {
      if (i === 0 || i === d.length) {
        return null
      }
      if (d[i] === '.' || d[i - 1] === '.' || d[i] === '-' || d[i - 1] === '-') {
        return null
      }
      return d.substr(0, i) + '.' + d.substr(i)
    }

  ).concat(

    // Transposition
    function (d, i) {
      if (d[i] === d[i + 1] || d[i + 1] === undefined) {
        return null
      }
      return d.substr(0, i) + d[i + 1] + d[i] + d.substr(i + 2)
    }

  ).concat(

    // Vowel swap
    range(vowels.length).map(function (v) {
      return function (d, i) {
        var vowelIndex = v % vowels.length

        if (vowels.indexOf(d[i]) === -1 || vowels[vowelIndex] === d[i]) {
          return null
        }

        return d.substr(0, i) + vowels[vowelIndex] + d.substr(i + 1)
      }
    })

  ).concat(

    // Miscellaneous
    range(4).map(function (m) {
      return function (d, i) {
        if (i === 0) {
          if (m === 0) {
            if (d.substring(0, 4) !== 'www.') {
              return 'ww' + d
            }
            return null
          } else if (m === 1) {
            if (d.substring(0, 4) !== 'www.') {
              return 'www' + d
            }
            return null
          } else if (m === 2) {
            if (d.substring(0, 4) !== 'www.') {
              return 'www-' + d
            }
            return null
          } else if (m === 3) {
            return d + 'com'
          }
        }
        return null
      }
    })

  )

  // Return the domain and TLD as two array elements.
  var split = function (domain) {
    var tld = tldjs.getPublicSuffix(domain)
    return [domain.slice(0, domain.length - tld.length - 1), tld]
  }

  var tweak = function (domain, cursor) {
    var parts = split(domain)

    var subdomains = parts[0]
    var tld = parts[1]

    var funcIndex = cursor % tweaks.length
    var tweakFunc = tweaks[funcIndex]
    if (tweakFunc === undefined) {
      return null
    }

    var charIndex = Math.floor(cursor / tweaks.length)
    if (charIndex > (subdomains.length - 1)) {
      return null
    }

    var tweakedDomain = tweakFunc(subdomains, charIndex)
    if (tweakedDomain === null) {
      return tweak(domain, cursor + 1)
    }

    return {
      'domain': tweakedDomain + '.' + tld,
      'cursor': cursor
    }
  }

  return {
    tweak: tweak
  }
})()
